lib/teracy-dev/plugin.rb

```diff
@@ -15,24 +15,38 @@ def self.sync(plugins)
 
       plugins.each do |plugin|
 
-        if !installed_plugins.has_key?(plugin['name']) and plugin['state'] == 'installed'
-          logger.info("installing plugin: #{plugin}")
-
-          if plugin['sources'].nil? or plugin['sources'].empty?
-            plugin['sources'] = [
-              "https://rubygems.org/",
-              "https://gems.hashicorp.com/"
-            ]
-          end
-
-          plugin_manager.install_plugin(plugin['name'], Util.symbolize(plugin))
-          reload_required = true
+        name_validate = validate(plugin, 'name' => plugin['name'])
+        if !name_validate
+          logger.warn("Plugin name must be configured, not nil, and not empty")
+          next
         end
 
-        if installed_plugins.has_key?(plugin['name']) and plugin['state'] == 'uninstalled'
-          logger.info("uninstalling plugin: #{plugin['name']}")
-          plugin_manager.uninstall_plugin(plugin['name'])
-          reload_required = true
+        case plugin['state']
+        when 'installed'
+          if installed_plugins.empty?
+            source_validate = validate(plugin, 'sources' => plugin['sources'])
+            if !source_validate
+              logger.warn("Plugin sources should be configured when no plugin installed, or we will use default")
+              plugin['sources'] = [
+                "https://rubygems.org/",
+                "https://gems.hashicorp.com/"
+              ]
+            end
+          end
+
+          if !installed_plugins.has_key?(plugin['name'])
+            logger.info("installing plugin: #{plugin}")
+            plugin_manager.install_plugin(plugin['name'], Util.symbolize(plugin))
+            reload_required = true
+          end
+        when 'uninstalled'
+          if installed_plugins.has_key?(plugin['name']) and plugin['state'] == 'uninstalled'
+            logger.info("uninstalling plugin: #{plugin['name']}")
+            plugin_manager.uninstall_plugin(plugin['name'])
+            reload_required = true
+          end
+        else
+          logger.debug('Plugin state not be set, do nothing')
```

Comment:
when plugin.enabled is `true` and `state` is not setted, we should fill it with `installed` state

Reply:
plugin state is not set, then do nothing is expected. It means that if the plugin is installed or not installed, users don't care.

Reply:
okey, by the way, the message should be "The plugin state is not set, no action will be taken."